macros/edr/data_monitoring/monitors_query/dimension_monitoring_query.sql

```diff
@@ -17,75 +25,93 @@
         dimensions_buckets as (
             select edr_bucket_start, edr_bucket_end, dimension_value
-            from buckets left join dimension_values_union on buckets.joiner = dimension_values_union.joiner
+            from unique_previous_dimension_values left join buckets
+                on (buckets.joiner = unique_previous_dimension_values.joiner
+                {# This makes sure we don't create empty buckets for dimensions before their first apperance #}
+                and edr_bucket_end >= dimension_min_bucket_end)
+            where dimension_value is not null
         ),
 
         {# Calculating the row count for the value of each dimension #}
-        filtered_row_count_values as (
-            select 
+        row_count_values as (
+            select
+                edr_bucket_start,
+                edr_bucket_end,
                 start_bucket_in_data,
-                {{ concat_dimensions_sql_expression }} as dimension_value, 
-                {{ elementary.edr_cast_as_float(elementary.row_count()) }} as row_count_value
-            from filtered_monitored_table
-            {{ dbt_utils.group_by(2) }}
+                dimension_value,
+                case when start_bucket_in_data is null then
+                    0
+                else {{ elementary.edr_cast_as_float(elementary.row_count()) }} end as row_count_value
+            from buckets left join time_filtered_monitored_table on (edr_bucket_start = start_bucket_in_data)
+            group by 1,2,3,4
         ),
 
         {# Merging between the row count and the dimensions buckets #}
-        {# This way we make sure that if a dimension has no rows in a day, it will get a metric with value 0 #}
-        row_count_values as (
-            select edr_bucket_start,
-                   edr_bucket_end,
+        {# This way we make sure that if a dimension has no rows, it will get a metric with value 0 #}
+        fill_empty_buckets_row_count_values as (
+            select dimensions_buckets.edr_bucket_start,
+                   dimensions_buckets.edr_bucket_end,
                    start_bucket_in_data,
                    dimensions_buckets.dimension_value,
                    case when start_bucket_in_data is null then
                        0
                    else row_count_value end as row_count_value
-            from dimensions_buckets left join filtered_row_count_values on (edr_bucket_start = start_bucket_in_data and dimensions_buckets.dimension_value = filtered_row_count_values.dimension_value)
+            from dimensions_buckets left join row_count_values
+                on (dimensions_buckets.edr_bucket_start = start_bucket_in_data and dimensions_buckets.dimension_value = row_count_values.dimension_value)
+        ),
+
+        union_row_count_values as (
+            select distinct *
+            from
+            (
+            select * from row_count_values
+              union all
+            select * from fill_empty_buckets_row_count_values
```

Comment:
`fill_empty_buckets_row_count_values` is built from `from dimensions_buckets left join row_count_values`.
Wouldn't union it with `row_count_values` is redundant and only causing  duplicate rows that we distinct them?
If not, please add a comment that explain this because it is hard to understand it frm reading the code 🙂

Reply:
For reference, on the no timestamp part, the `fill_empty_buckets_row_count_values` is really only the filled values

Reply:
We union so new buckets added in this run will be included (were filtered by the join we did on 'fill_empty_buckets_row_count_values'). I added a comment